### PR TITLE
Work on the introspection API

### PR DESCRIPTION
**Pull request by pmderodat**

### CODE REVIEW

File: langkit/templates/pkg_introspection_body_ada.mako [hunk 1]
```diff
@@ -268,27 +346,94 @@ package body ${ada_lib_name}.Introspection is
...
+      % else:
+         ${return_program_error()}
       % endif
+   end Fields;
 
+   ------------
+   -- Fields --
+   ------------
+
+   function Fields
```
Comment (raph-amiard) [inline langkit/templates/pkg_introspection_body_ada.mako]: This looks like expensive work that could be cached no ?

File: langkit/templates/pkg_introspection_body_ada.mako [hunk 2]
```diff
@@ -9,11 +9,38 @@ with Ada.Strings.Unbounded.Hash;
 with ${ada_lib_name}.Implementation;
```
Comment (raph-amiard) [inline langkit/templates/pkg_introspection_body_ada.mako]: General comment: this commit seems to be a bit doc-lacking. There are both undocumented types and IMHO the code generation part could be a bit more documented.

File: langkit/templates/pkg_introspection_spec_ada.mako [hunk 3]
```diff
@@ -68,21 +68,37 @@ package ${ada_lib_name}.Introspection is
    function Derived_Types (Id : Node_Type_Id) return Node_Type_Id_Array;
    --  Return type references for all direct derivations for Id
 
+   <% all_abstract = ctx.sorted_parse_fields + ctx.sorted_properties %>
+
+   type Abstract_Node_Data_Reference is
```
Comment (raph-amiard) [inline langkit/templates/pkg_introspection_spec_ada.mako]: While this is the Langkit name, maybe `Node_Data_Reference` is sufficient for the public API name, what do you think ?

File: langkit/templates/pkg_introspection_spec_ada.mako [hunk 4]
```diff
@@ -68,6 +68,12 @@ package ${ada_lib_name}.Introspection is
    function Derived_Types (Id : Node_Type_Id) return Node_Type_Id_Array;
    --  Return type references for all direct derivations for Id
 
+   function Is_Derived_From (Id, Parent : Node_Type_Id) return Boolean;
+   --  Return whether true if at least one of the following is true:
```
Comment (raph-amiard) [inline langkit/templates/pkg_introspection_spec_ada.mako]: This comment made my head hurt. This is the definition of the function but including the recursive definition in the comment is not very helpful IMHO :) Saying it returns true if the type identified by Id is either equal to parent, directly or indirectly derived from parent would seem better

File: langkit/templates/pkg_introspection_body_ada.mako (hunk 1 above)
Comment (pmderodat) [reply langkit/templates/pkg_introspection_body_ada.mako]: As discussed live: this algorithm runs in linear time for number of fields + number of derivations. Caching is possible, but I would suggest to wait for reported performance issue to bother introducing a cache. :-)

File: langkit/templates/pkg_introspection_body_ada.mako (hunk 2 above)
Comment (pmderodat) [reply langkit/templates/pkg_introspection_body_ada.mako]: After live discussion: this note was written thinking this patch hunk was for a package specification (which contains most documentation). This is the body, so the API is not documented here on purpose, but internals should be discribed as needed. We’ll add more comments if necessary.

File: langkit/templates/pkg_introspection_spec_ada.mako (hunk 3 above)
Comment (pmderodat) [reply langkit/templates/pkg_introspection_spec_ada.mako]: Good point, I’ll rename, thanks!

File: langkit/templates/pkg_introspection_spec_ada.mako (hunk 4 above)
Comment (pmderodat) [reply langkit/templates/pkg_introspection_spec_ada.mako]: Right, sorry, the comment describes the computation more than what it expresses. I’ll fix that.

**Pull request merged**